### ISSUE TITLE
Different results from max pooling layer when using theano and tensorflow

### ISSUE THREAD

**Issue by piergiaj** [fchollet/keras#5090]
I'm working with the SSD model (https://github.com/rykov8/ssd_keras) and it works fine in tensorflow. I've converted all the tensorflow parts to theano, including the conv kernels (as described here: https://github.com/fchollet/keras/wiki/Converting-convolution-kernels-from-Theano-to-TensorFlow-and-vice-versa) and the model works almost perfectly. 

However, I've run into an issue where the output from the max pooling layer when running tensorflow is different from the output in theano. It is a very small difference, but it moves the bounding boxes slightly.

For example, the outputs from conv1_1 and conv1_2 are identical, but the outputs from pool1 in tensorflow give (just a small sample of the output):

```
 [  7.90634384e+01   0.00000000e+00   0.00000000e+00 ...,
      7.58856506e+01   2.82682953e+02   4.01717926e+02]
   [  9.10728989e+01   0.00000000e+00   0.00000000e+00 ...,
      8.20418091e+01   3.42485565e+02   3.65100067e+02]
   [  1.60205139e+02   0.00000000e+00   0.00000000e+00 ...,
      7.80331039e+01   4.03439850e+02   3.21558014e+02]
```

while the outputs in theano give
```
[  6.89849014e+01   1.28302107e+01   0.00000000e+00 ...,
      7.81133041e+01   3.79435974e+02   4.01718109e+02]
   [  8.10280304e+01   2.65449944e+01   0.00000000e+00 ...,
      7.58856430e+01   3.21112823e+02   3.87413025e+02]
   [  2.77083099e+02   0.00000000e+00   0.00000000e+00 ...,
      8.20418396e+01   3.40778381e+02   3.30199341e+02]
```

Is there some difference in the way keras, theano or tensorflow handle 2D max pooling? I would expect the results to be identical, but they are not.

Here is the model I'm using to test this:
```
from keras.layers import Activation
from keras.layers import Convolution2D
from keras.layers import Input
from keras.layers import MaxPooling2D
from keras.models import Model



def TestModel(input_shape, num_classes=21):
    net = {}
    # Block 1                                                                                                                                                                                                
    input_tensor = input_tensor = Input(shape=input_shape)
    img_size = (input_shape[1], input_shape[0])
    net['input'] = input_tensor
    net['conv1_1'] = Convolution2D(64, 3, 3,
                                   activation='relu',
                                   border_mode='same',
                                   name='conv1_1')(net['input'])
    net['conv1_2'] = Convolution2D(64, 3, 3,
                                   activation='relu',
                                   border_mode='same',
                                   name='conv1_2')(net['conv1_1'])
    net['pool1'] = MaxPooling2D((2, 2), strides=(2, 2), border_mode='same',
                                name='pool1')(net['conv1_2'])

```

The weights I downloaded from [here](https://mega.nz/#F!7RowVLCL!q3cEVRK9jyOSB9el3SssIA).

**Comment by piergiaj** [comment]
I found the problem, I changed the lines [1749-1750](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L1749-L1755) in the theano_backend file from
```
    if border_mode == 'same':
        expected_width = (x.shape[2] + strides[0] - 1) // strides[0]
        expected_height = (x.shape[3] + strides[1] - 1) // strides[1]

        pool_out = pool_out[:, :,
                            : expected_width,
                            : expected_height]

```

to:
```
    if border_mode == 'same':
        expected_width = (x.shape[2] + strides[0] - 1) // strides[0]
        expected_height = (x.shape[3] + strides[1] - 1) // strides[1]
        height = pooled_out.shape[3]
        width = pooled_out.shape[2]

        pool_out = pool_out[:, :,
                            width - expected_width :,
                            height - expected_height :]

```

I'm not sure why that works, and it would probably break many existing models, but it seems like pool2d with 'SAME' padding in tensorflow implements it this way.

**Comment by fchollet** [comment]
Thanks, that's good to know.

**Comment by pasky** [comment]
Would this seem reasonable to fix, @fchollet? (Not rushing to make a PR as I don't really use theano anymore but this confused the heck out of me when running a short test on a theano-based machine.)

Simple way to reproduce - the script http://paste.debian.net/914226/ returns the matrix shown in the script with theano, rather than the correct matrix:
```
[[ 17.  18.  19.  20.  21.  21.  21.]
 [ 24.  25.  26.  27.  28.  28.  28.]
 [ 31.  32.  33.  34.  35.  35.  35.]
 [ 38.  39.  40.  41.  42.  42.  42.]
 [ 45.  46.  47.  48.  49.  49.  49.]
 [ 45.  46.  47.  48.  49.  49.  49.]
 [ 45.  46.  47.  48.  49.  49.  49.]]
```

I don't actually understand the magic w_pad calculation in the code referenced in previous comment, why isn't it just `w_pad = pool_size[0] // 2` ?  If I ZeroPadding2D manually like that and use `border_mode='valid'`, I get the correct matrix with both backends.

So that's the workaround right now.  But esp. for people doing pyramid pooling and other scenarios with huge pool sizes, pooling is broken for them with theano without them probably even aware of it.

**Comment by piergiaj** [comment]
I don't know if it should be changed, as it may break existing trained theano models that learned using this (possibly) incorrect pooling, which is why I didn't submit a pull request, but simply left the solution here for others who encounter this problem.

I've also used the ZeroPadding and 'valid' border mode with both back ends sucessfully, but the SSD model I was converting used 'same' border mode, with led to this error.

The reason it isn't simply pool_size // 2 is because this allows strided pooling. So if the stride is 1, there is no reduction in size, or larger strides lead to more reduction in size. That equation follows the same as the convolution output sizing. The difference I found is that the current theano removes excess numbers from the bottom right of the "image" while tensorflow removes from the top left. Not sure why without looking into the implementations, but this fixed it.

**Comment by piergiaj** [comment]
I further tested this today, converting a caffe model to theano directly. Interestingly, the way caffe pooling with the same/half border mode works is 

```    if border_mode == 'same':
        expected_width = (x.shape[2] + strides[0] - 1) // strides[0]
        expected_height = (x.shape[3] + strides[1] - 1) // strides[1]

        pool_out = pool_out[:, :,
                            : expected_width,
                            : expected_height]
```

which is opposite of tensorflow pooling. I don't know the best way to handle this, as theano, tensorflow and caffe all seem to have slightly different ways to do pooling.

**Comment by BlaiMelendezCatalan** [comment]
Just in case this is useful to anybody, the behaviour of a TF MaxPooling2D with 'SAME' padding can be achieved using a theano backend:

1) Place a ZeroPadding layer with the necessary padding (probably something like padding=((0, 1), (0, 1)) ) before the MaxPooling2D layer.
2) Set the MaxPooling2D layer padding to 'VALID'.
